Display misses in the hit accuracy heatmap

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu/pull/27062 (because it looks wrong otherwise)

These were always intended to be displayed, just implementation was lacking. I've changed them to crosses as I feel it fits better than circles.

Examples:

<img width="1624" alt="Screenshot 2024-02-06 at 22 23 37" src="https://github.com/ppy/osu/assets/1329837/4be948d8-b558-476d-82a7-2b92d27a70e4">
<img width="1624" alt="Screenshot 2024-02-06 at 22 25 14" src="https://github.com/ppy/osu/assets/1329837/a1d1d060-d479-4725-aaae-a0544d3c5e91">
<img width="1624" alt="Screenshot 2024-02-06 at 22 25 50" src="https://github.com/ppy/osu/assets/1329837/bdb5dbdf-768e-4884-9d50-311847fcb9d0">

smoogipoo committed:
Show near-misses on the results-screen heatmap

smoogipoo committed:
Make miss points into crosses

bdach committed:
Merge branch 'master' into heatmap-misses

bdach commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -264,12 +259,31 @@ public HitReceptor()
 
             public bool OnPressed(KeyBindingPressEvent<OsuAction> e)
             {
+                if (!(CanBeHit?.Invoke() ?? false))
```
Good lord this is convoluted. I have to write down a truth table to even figure out what this is trying to compute.

Can we simplify to

```suggestion
                if (CanBeHit?.Invoke() != true)
```

(or at least I *think* that's the equivalent...)

smoogipoo replied:
I've taken a bit of a different path, by enabling NRT for this file.

In the process, I've taken liberty of using a little bit of a newer syntax: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/required

I hope this is okay and agreeable (I think it can be used in more places). The alternative would be to use the ctor to indicate that it's required, but imo this is a _teeny_ bit cleaner.

bdach replied:
I could be fine with `required` properties as long as @peppy is. I'm personally a fan.

bdach commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -247,9 +240,11 @@ public partial class HitReceptor : CompositeDrawable, IKeyBindingHandler<OsuActi
             public override bool HandlePositionalInput => true;
 
-            public Func<bool> Hit;
+            public Func<bool> CanBeHit;
+            public Action Hit;
```
I think these two require xmldoc at this point...

smoogipoo committed:
Remove LangVersion redefintions

smoogipoo committed:
Enable NRT for DrawableHitCircle to clean up

smoogipoo committed:
Add xmldocs

smoogipoo committed:
Fix inverted condition

smoogipoo committed:
Fix inspections

smoogipoo committed:
Always process position as long as it's hittable

For example... If a hitobject is pressed but the result is a shake, this
will stop processing hits.

cdwcgt commented:
Can we also record the time we hit so that we can show miss in bar hit meter?
![image](https://github.com/ppy/osu/assets/34775378/6ebe7a83-a2ac-42bc-ba51-c725c3b2dd26)

bdach committed:
Merge branch 'master' into heatmap-misses

bdach commented on osu.Game.Rulesets.Osu/Statistics/AccuracyHeatmap.cs:
```diff
@@ -250,48 +257,73 @@ protected void AddPoint(Vector2 start, Vector2 end, Vector2 hitPoint, float radi
+                return;
 
-            PeakValue = Math.Max(PeakValue, ((HitPoint)pointGrid.Content[r][c]).Increment());
+            PeakValue = Math.Max(PeakValue, ((GridPoint)pointGrid.Content[r][c]).Increment());
```
So on another look I'm not entirely sure this does what it's supposed to be doing?

Why? Because you can _be inside the hitcircle and still miss_ if you hit the circle too early.

On this branch, the following can happen:

https://github.com/ppy/osu/assets/20418176/0ff6ced3-67fa-4e2e-b31b-c32be1e3d6d4

I missed all three circles due to hitting too early but did not fail because HP didn't drain enough. The accuracy heatmap is expectedly empty. On the heatmap, two hits are registered, even though _I didn't actually hit anything_.

I'm not sure what to do with this or if this is undesirable even, but it does seem counterintuitive at least. I would probably just not count such hits.

smoogipoo replied:
Personally I think that's fine.

bdach reviewed (approved):
i'm just gonna get this in, it's waited long enough
